Add `Env`, `Fresh` and `GetOpt` effects

just95 opened the pull request:
<!--
  Have you read our Code of Conduct?
  By filing an issue or pull request, you are expected to comply with it, including treating everyone with respect:
  https://github.com/FreeProving/guidelines/blob/main/CODE_OF_CONDUCT.md
-->

### Issue

<!--
  Link to the issue that this pull request addresses.
  If there is not yet a corresponding issue, please open a new issue and then link to that issue in your pull request.
  Give any additional information that is not covered by the linked issue but might be important for the reviewer.
-->
Closes #31.

### Description of the Change

<!--
  Give a short summary of the changes you made to handle the issue.
  Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.
-->
The `PM` state monad has been replaced by three Polysemy effects.

 - `Env` is a state effect that provides the pattern matching compiler's `Environment` to computations. The `Environment` stores constructors that are in scope (needed for case completion) as well a which pattern variables have been matched against (needed for optmizations).
 - `Fresh` is an effect for the generation of fresh variables. `Fresh` is more flexible than `PM` since it allows other prefixes than `a` for fresh variables. However, all generated fresh variables still use the prefix `a` at the moment. Similar to `PM`, handlers for the `Fresh` effect are not aware of functions and local variables. Thus, the new effect is also affected by #15.
 - `GetOpt` is a reader effect for the command line `Options`. This effect required me to move the `Options` data type from `Main` to its own module in `./src/lib/`.

just95 committed:
Move `Options` to own module #31

just95 committed:
Add reader effect for command line options #31

just95 committed:
Add effect for generation og fresh variables #31

just95 committed:
Add dummy `Environment` type #31

just95 committed:
Add effect to read and write environment #31

just95 committed:
Add constructor entries to `Environment` #31

just95 committed:
Add matched patterns to `Environment` #31

just95 committed:
Fix initialization of empty environment #31

just95 committed:
Remove unused code #31

just95 committed:
Add missing comments #31

just95 committed:
Format TODO comments

just95 committed:
Add entries for data types to `Environment` #31

just95 committed:
Add lookup and insert for `Environment` #31

just95 committed:
Add `GetOpt` handler that parses arguments #31

just95 committed:
Run handlers for `Env` and `Fresh` in `Main` #31

just95 committed:
Merge branch 'issue-25' into issue-31

just95 committed:
Remove redundant imports #31

just95 committed:
Remove redundant dependency #31

just95 committed:
Tidy up `Application` module #31

just95 committed:
Derive `Ord` instance for names #31

In order to use the names of our intermediate syntax
in `Map`s, we have to define an `Ord` instance.
An order has been defined on source spans that treats
all source spans as equal.

just95 committed:
Parameterize `Environment` #31

The environment contains names and patterns. Since names and
patterns are paremterized over type families for additional
information, the environment has to be parameterized as well
or the information must be discarded before a node is inserted
into the environment and restored when it is looked up.

just95 committed:
Initialize `Env` in `Application` #31

just95 committed:
Add environment entries for built-in types #31

just95 committed:
Initialize built-in constructors in `Application` #31

just95 committed:
Fix tests for `Application` #31

just95 committed:
Simplify `Decl` type #31

Type signatures are not needed by the pattern matching compiler.
Thus, the `TypeSig` constructor of `Decl` could be removed.
Furthermore, the name of the data type is the only information
from the data type declarations's head. Thus, the `DeclHead`
constructor could be removed in favor of a `Name` field in
for the `DataDecl` constructor.

just95 committed:
Remove `RecDecl` constructor from `ConDecl` #31

Record constructor declarations are not supported.
They were used only to throw an error in `makeConEntry`.
The error is now thrown during the conversion to the
intermediate representation already.

just95 committed:
Add convenience actions for `Fresh` #31

Currently there is only an action to generate an indentifier
for a fresh variable.
However, it is often necessary to generate fresh varibale
expressions or patterns. Additional actions with fixed
interpretations have been added to generate such AST nodes.

just95 committed:
Initialize `nextId` in `Application` #31

just95 committed:
Use `Fresh` in guard elimination #31

just95 committed:
Use effects in core algorithm #31

just95 committed:
Remove `HST.Environment.FreshVars` #31

just95 committed:
Remove redundant `S.EqAST` constraints #31

just95 committed:
Add missing comment #31

just95 committed:
Apply HLint suggestions #31

just95 committed:
Format code #31

just95 committed:
Fix generation of Haddock documentation #31

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -3,82 +3,116 @@
+  insertPreludeEntries
+  collectDataInfo m
+  eliminatedM    <- applyGEModule m
+  caseCompletedM <- applyCCModule eliminatedM
```
```suggestion
  guardEliminatedM <- applyGEModule m
  caseCompletedM   <- applyCCModule guardEliminatedM
```

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -3,82 +3,116 @@
+--   matches.
+--
+--   If the function has only one rule and no pattern is a constructor
+--   pattern, the algorithm is is left unchanged.
```
```suggestion
--   pattern, the function is is left unchanged.
```

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -89,73 +123,42 @@ useAlgo ms = do
-    )    -- TODO Tuples
-  , ("wildcard", [(S.Special S.NoSrcSpan (S.ExprHole S.NoSrcSpan), 0, False)])
-  ]
+-- | Creates an environment entry for a constructor declaration .
```
```suggestion
-- | Creates an environment entry for a constructor declaration.
```

stu204767 commented on src/lib/HST/CoreAlgorithm.hs:
```diff
@@ -232,19 +242,28 @@ getQNamePat _ = error "getQNamePat unsuported Pattern"
+  :: (Member Fresh r, S.EqAST a)
   => S.Pat a         -- ^ The fresh variable matched by the @case@ expression.
-  -> [Constructor a] -- ^ The missing constructors to generate alternatives for.
+  -> [ConEntry a] -- ^ The missing constructors to generate alternatives for.
```
```suggestion
  -> [ConEntry a]    -- ^ The missing constructors to generate alternatives for.
```

stu204767 commented on src/lib/HST/Effect/Fresh.hs:
```diff
@@ -0,0 +1,90 @@
+-- Effect and Actions                                                        --
+-------------------------------------------------------------------------------
+
+-- | An effect capable of generating names for a fresh variable.
```
This should both be in singular or both be in plural:
```suggestion
-- | An effect capable of generating names for fresh variables.
```

stu204767 commented on src/lib/HST/Environment/Prelude.hs:
```diff
@@ -0,0 +1,102 @@
+                                                )
+import qualified HST.Frontend.Syntax           as S
+
+-- | Inserts entries for build-in data types and data constructors into the
```
```suggestion
-- | Inserts entries for built-in data types and data constructors into the
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -27,12 +33,13 @@ type GExp a = ([S.Pat a], S.Rhs a)
+  :: Members '[Fresh] r
+  => [S.Pat a]  -- fresh Vars
   -> S.Exp a  -- error
   -> [GExp a] -- pairs of pattern and guarded rhs
```
```suggestion
  => [S.Pat a] -- fresh Vars
  -> S.Exp a   -- error
  -> [GExp a]  -- pairs of pattern and guarded rhs
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -47,21 +54,23 @@ toDecl _ _ = error "GuardElimination.toDecl: Variable pattern expected"
+  :: Member Fresh r
+  => [S.Pat a]                 -- fresh variables for the case exps
   -> ([S.Decl a], S.Pat a) -- startcase ([], first generated Pattern)
   -> [GExp a]                -- list of pattern + rhs pair
```
```suggestion
  => [S.Pat a]                   -- fresh variables for the case exps
  -> ([S.Decl a], S.Pat a)       -- startcase ([], first generated Pattern)
  -> [GExp a]                    -- list of pattern + rhs pair
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -47,21 +54,23 @@ toDecl _ _ = error "GuardElimination.toDecl: Variable pattern expected"
+  :: Member Fresh r
+  => [S.Pat a]                 -- generated varibles
   -> ([S.Decl a], S.Pat a) -- (current decls , variable for let binding)
   -> GExp a                  -- pairs of pattern to match against and a guarded Rhs
```
```suggestion
  => [S.Pat a]                   -- generated variables
  -> ([S.Decl a], S.Pat a)       -- (current decls , variable for let binding)
  -> GExp a                      -- pairs of pattern to match against and a guarded Rhs
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -87,24 +96,25 @@ createCase i next ((v, p) : vps) = S.Case
-  :: S.Rhs a      -- the (maybe guarded) righthandside
+  :: Member Fresh r
+  => S.Rhs a      -- the (maybe guarded) righthandside
   -> S.Exp a      -- next case
```
```suggestion
  => S.Rhs a         -- the (maybe guarded) right hand side
  -> S.Exp a         -- next case
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -87,24 +96,25 @@ createCase i next ((v, p) : vps) = S.Case
-  buildIF
+  buildIf
     :: S.Exp a               -- next rule
     -> [S.GuardedRhs a]      -- guarded rhs to fold
```
```suggestion
    :: S.Exp a          -- next rule
    -> [S.GuardedRhs a] -- guarded rhs to fold
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -87,24 +96,25 @@ createCase i next ((v, p) : vps) = S.Case
     -> [S.GuardedRhs a]      -- guarded rhs to fold
-    -> PM a (S.Exp a)    -- if then else expr
-  buildIF nx gs = foldM
+    -> Sem r (S.Exp a)    -- if then else expr
```
```suggestion
    -> Sem r (S.Exp a)  -- if then else expr
```

stu204767 commented on src/lib/HST/Frontend/FromHSE.hs:
```diff
@@ -55,11 +53,11 @@ transformDecl (HSE.PatBind s (HSE.PVar _ name) rhs mBinds) = Just
-transformDeclHead (HSE.DHApp _ dHead _  ) = S.DHApp (transformDeclHead dHead)
+transformDeclHead :: HSE.DeclHead HSE.SrcSpanInfo -> S.Name HSE
+transformDeclHead (HSE.DHead _ dName    ) = transformName dName
+transformDeclHead (HSE.DHInfix _ _ dName) = transformName dName
```
An error was previously thrown explicitely for infix data declaration heads in the `Application` module. Are those data declarations fully supported now?

just95 replied:
They are [not supported by the Free Compiler](https://github.com/FreeProving/free-compiler/blob/main/src/lib/FreeC/Frontend/Haskell/Simplifier.hs#L367) at the moment but I don't think that there is a reason not to allow them.

just95 replied:
See also <https://github.com/FreeProving/free-compiler/pull/139>.

stu204767 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -47,41 +55,42 @@ newtype Module a = Module [Decl a]
 --   represented by function bindings.
-data Decl a = DataDecl (DeclHead a) [ConDecl a]
-            | TypeSig (SrcSpan a) [Name a] (TypeExp a)
+data Decl a = DataDecl (Name a) [ConDecl a]
             | FunBind (SrcSpan a) [Match a]
```
The reason I implemented the `TypeSig` constructor was to support and retain type signatures in `where` bindings, as these are quite common. In my opinion, this is reason enough to leave the `TypeSig` constructor inserted.
Or do you have a different idea for those type signatures?

just95 replied:
Ah okay, that makes sense. I didn't think about `where` clauses and `let` expressions. However, as far as I know `where` clauses are not supported anyway and will even be discarded silently at the moment. Thus, we should probably remove them from the intermediate syntax and print an error message if there is a `where` clause.
In `let` expressions it would still be nice to preserve type signatures. Maybe we can add a more general `UnsupportedDecl` constructor for this purpose. This constructor would also allow us to remove the input module parameter from `ToHSE.transformModule`.

just95 replied:
See 60574d0f9ba616824bb6b22a7cf7475a8a169495.

stu204767 committed:
Apply simple suggestions from my own code review

just95 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -97,16 +97,16 @@ createCase i next ((v, p) : vps) = S.Case
   :: Member Fresh r
-  => S.Rhs a      -- the (maybe guarded) righthandside
-  -> S.Exp a      -- next case
+  => S.Rhs a         -- the (maybe guarded) right hand side
```
```suggestion
  => S.Rhs a         -- the (maybe guarded) right-hand side
```

just95 committed:
Update src/lib/HST/Feature/GuardElimination.hs

just95 committed:
Update src/lib/HST/Effect/Fresh.hs

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

just95 committed:
Update src/lib/HST/Application.hs

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

just95 committed:
Add `S.TypeSig` back #31

Type signatures where removed in 123cee48856938698326ccfbd25fa13af89b4b93
in order to simplify the `S.Decl` data type. As was pointed out in
https://github.com/FreeProving/haskell-src-transformations/pull/37#discussion_r455002501
type signature AST nodes are required to preserve type signatures
in `let` expressions.
